### ISSUE TITLE
Placing Variables on the cpu using `tf.contrib.layers` functions

### ISSUE THREAD

**Issue by jhaux** [tensorflow/tensorflow#9517]
Dear tensorflow team,

After constructing my model using the functionality provided by `tf.contrib.layers` I now want to extend my model over several GPUs. I learned that it might be beneficary to place Variables on the CPU when doing that, to reduce data transfer overhead. After not seeing an easy way to do this I found a workaround I described on [stackoverflow](http://stackoverflow.com/questions/43678599/device-placement-of-kernels-and-biases-when-using-tf-contrib-layers/43685023#43685023). My solution is to generate Variable-nodes in the graph where the Variable-getter of the `fully_connected` layer for example would expect the variables to be.

As this is not a very nice solution, I messed with the `fully_connected` layer and the `_build_variable_getter` function to basically allow me to specify, where I want to place the variabels. Thus after 

- adding the kwarg `variable_device` to `tf.contrib.layer.fully_connected`
- adding the kwarg `variable_device` to `tf.contrib.layer._build_variable_getter`
- adding the kwarg `device` to `tf.contrib.layer._model_variable_getter`
- and passing this as kwarg to `model_variable` defined in `tensorflow.contrib.framework.python.ops.variables`

I get the desired functionality when using the `fully_connected` layer.
Below you find the modified version of `layers.py` 

In my eyes this would be a very useful feature for all layers that contain trainable variables, which is why I would like to make a request for this feature.
If you think the supplied modification is good enough, I can also try to make a pull request, after updating the other layers.

Best, Johannes

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: As described above, I did
- **OS Platform and Distribution**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version**: ('v1.0.0-65-g4763edf-dirty', '1.0.1')
- **CUDA/cuDNN version**: 8, 5.1
- **GPU model and memory**: Titan X Pascal, 12GB

### Source code / logs
See 
- [stackoverflow](http://stackoverflow.com/questions/43678599/device-placement-of-kernels-and-biases-when-using-tf-contrib-layers/43685023#43685023)
- [modified layers.py](https://drive.google.com/file/d/0BxImZIuERGB2MHBMRDZkejM4LVU/view?usp=sharing)

**Comment by aselle** [comment]
@tfboyd, didn't you run into this as well? @martinwicke, do you have any api design thoughts on this?

**Comment by tfboyd** [comment]
I meant to ask for this.  @martinwicke and I spoke about it but I never formally asked.  This would be a good add to be able to state where you want the variables to go.  Just this type of change should be a 10-12% speedup using the current stock slim models as tested with resnet-50 a couple months ago.  Putting the variables on cpu has been optimal for many models, e.g. inception and resnet, regardless of GPU.  It even works great for Tesla P100s on systems with high speed GPU links.  Putting shared variables on GPU:0 has only been effective in a limited number of situations and architectures that I have tested, e.g. AlexNet on AWS K80s, and VGG16 again only on AWS K80s.

**Comment by martinwicke** [comment]
The canonical way to do this is to use a device function.

Or variable_scope, but I think we should encourage device functions.

**Comment by jhaux** [comment]
Could you elaborate on this @martinwicke? Or guide me to the documentation? 
I understand how to place variables using context managers, but in this case I would like the Variable nodes to be created by the layer function.

E.g. it would be desirable for me to be able do something like this:
`with tf.device('\gpu:0):
    fc = fully_connected (input, ..., variable_device='\cpu:0')`
Here the variables would be created on the CPU while the matrix multiplication and bias add are done on the GPU.

**Comment by jhaux** [comment]
Hey everybody,

Any news on this topic? 
I would be happy to supply a pull request, but I am not sure, my extensions of the code meet possible guidelines. Also I am not sure if my testing was sufficient. If you want you can have a look at [my fork of the tensorflow repository](https://github.com/jhaux/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py), where the changes can be found in all layers and getters, where I thought the `variable_device` keyword might make sense, e.g. where variables are created or fetched. Those include:

- `bias_add`
- `convolution`
- `convolution2d_in_plane`
- `convolution2d_transpose`
- `_model_variable_getter`
- `_build_variable_getter`
- `fully_connected`
- `separable_convolution2d`

The changes should be backward compatible as only kwargs with default values = `None` are used.
I also added Documentation in the docstring:
`variable_device: Device where the variables of the layer (e.g. weights and biases) should be placed, e.g. /cpu:0`

I think this approach is more intuitive than using a `variable_scope` with a `custom getter`.

**Comment by martinwicke** [comment]
@tfboyd you were working on a device function for this exact purpose, can you post it here?

**Comment by leo-p** [comment]
Any news on this issue? It would definitely be a useful feature!

**Comment by tfboyd** [comment]
I will try to post example code tonight. It is not super complicated.  For image models (only thing I work with right now) CPU vs. GPU for the parameters can depend on the model and situation.  I will add a little more detail when I post the code sample.

**Comment by tfboyd** [comment]
Here we go.  I have actually changed my mind and agree with Martin.  Although I am not an authority as you will see from my sloppy python.  This is how I do it and it has worked for me in a few instances.  In fact using this with the SLIM models can boost performance on systems where GPUs are not peered or resnet which seems to often just work better with CPU:0 as the ps-server.  This is just how I do it and I am very open to criticism and a better approach.  I am still learning Python.

```python
tf.app.flags.DEFINE_string('local_ps_device', 'CPU:0', """Local parameter server GPU:0 if gpus are peered or CPU:0 otherwise try both.""")

PS_OPS = [
      'Variable', 'VariableV2', 'AutoReloadVariable', 'MutableHashTable',
      'MutableHashTableOfTensors', 'MutableDenseHashTable'
]

def assign_to_device(device, ps_device=None):
    """Returns a function to place variables on the ps_device.

    Args:
        device: Device for everything but variables
        ps_device: Device to put the variables on.  Example values are GPU:0 and
        CPU:0.
    
    If ps_device is not set then the variables will be placed on the device.
    The best device for shared varibles depends on the platform as well as the
    model.  Start with CPU:0 and then test GPU:0 to see if there is an 
    improvement.  

    """
    def _assign(op):
        node_def = op if isinstance(op, tf.NodeDef) else op.node_def
        if node_def.op in PS_OPS:
            return "/" + ps_device
        else:
            return device
    return _assign

def train():
       """ Incomplete example of a train method just to show the use of assign_to_device
       
       Note: The is from a ResNet and AlexNet training CIFAR-10 example on Multi-GPU. 
       I suspect there is some code "smell" but I was working with something that already 
       existed and the purpose is to illustrate the general usage.  Feedback welcome.  
       """
        tower_grads = []
        reuse_variables = None
        losses = []
        for i in six.moves.range(FLAGS.num_gpus):
             with tf.device(assign_to_device(
                            '/gpu:{}'.format(i),
                            ps_device=FLAGS.local_ps_device)):
                with tf.name_scope('{}_{}'.format('TOWER', i)) as n_scope:
                    with tf.device('/cpu:0'):
                        images, labels = cifar10_input.inputs(False, FLAGS.data_dir, FLAGS.batch_size)
                    with tf.variable_scope(tf.get_variable_scope(), reuse=reuse_variables):
                        logits = inference_small(images, is_training=True, num_blocks=9)
                    tower_loss = loss(logits, labels)
                    losses.append(tower_loss)
                    grads = optimizer.compute_gradients(tower_loss)
                    tower_grads.append(grads)
                    reuse_variables = True
....

```

**Comment by tfboyd** [comment]
I found that if the GPUs are not peered then CPU:0 is the often the best option.  If the GPUs are peered then GPU:0 was the best option for VGG16 and Alexnet.  InceptionV3 and Resnet were better on CPU:0 and even faster with replicated on GPU without NCCL.  For example: InceptionV3 was 214 imgs/sec on CPU vs. 218 images/sec using replicated with 8 GPUs batch-size 32 per GPU).  Maybe I am conservative but I wouldn't mess with replicated for an extra 4 images/second.  But if you are interested check out the [document](https://www.tensorflow.org/performance/performance_models) and sample code.  Replicated makes a much bigger difference for image models with more parameters (like VGG16) when using Tesla P100s.  CPU:0 is still the best choice (as of the time I wrote this message) for InceptionV3 and ResNet on P100s and replicated using NCCL for VGG16 and AlexNet.  

Let me know if this function closes out your request and we will close the issue.  I want to wait for your response before closing.  I expect there is may be some more valuable discussion / feedback.

**Comment by leo-p** [comment]
It's not exactly the same but I've found that using SLIM one can simply do:
```
with tf.device('/gpu:%d' % gpu):
    with tf.name_scope('tower_%d' % gpu) as scope:
        with slim.arg_scope([slim.model_variable, slim.variable], device='/cpu:0'):
            net = slim.fully_connected(...)
```

Which will pin the variables on `/cpu:0` while maintaining the actual computations on `/gpu:x`.

And don't forget to set `colocate_gradients_with_ops=True` when computing the gradients.

**Comment by tfboyd** [comment]
I learned and noticed something new today.  In the tf_cnn_benchmark code when we place the variables on the GPUs we spread them out even in the basic parameter server mode.  Here is a [link to the code](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/variable_mgr.py#L81) (long term the line number may change but for now the link goes to the correct function). This will get rolled into a much nicer example and I hope utility methods in the next 60 days or so.  

Personally, I am curious of the gain.  I hope to give this a test on a model I know works better with GPU as a parameter server and test GPU:0 vs. this method.  I hope this helps.  There are so many ways to approach these problems and variables that determine which is the best choice for each situation.

**Comment by JZDSS** [comment]
```
def net(x):
	res = tf.contrib.layers.conv2d(x, 10, [3, 3])
	...
	return res

with tf.device('/CPU:0'):
	with tf.name_scope('CPU'):
		net(x)

for i in range(num_gpus):
	with tf.device('/GPU:%d' % i):
		with tf.name_scope('GPU_%d' % i):
			with tf.variable_scope(tf.get_variable_scope(), reuse=True):
				y = net(x)
				loss = ...
				grad = ...
```
It works!
[http://jzdss.club/2018/03/04/Multi-GPUs/](http://jzdss.club/2018/03/04/Multi-GPUs/)

**Comment by yuke93** [comment]
@JZDSS Thanks! I believe this is a simple practical implementation :thumbsup:
